### ISSUE TITLE
Bazel build does not pick up correct compiler include paths

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#33975]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): RHEL 7.5
- TensorFlow installed from (source or binary): source
- TensorFlow version: 2.0.0
- Python version: 3.7.5
- Installed using virtualenv? pip? conda?: no
- Bazel version (if compiling from source): 0.26.1
- GCC/Compiler version (if compiling from source): GCC 8.2.0
- CUDA/cuDNN version: 10.1

**Describe the problem**

Bazel does not pick up the correct include paths of GCC and returns errors such as:

```
ERROR: /tmp/easybuild-tmp/eb-5QGVSJ/tmpMZolLj-bazel-build/external/fft2d/BUILD.bazel:27:1: undeclared inclusion(s) in rule '@fft2d//:fft2d':
this rule is missing dependency declarations for the following files included by 'external/fft2d/fft2d/fftsg2d.c':
  '/sw/installed/GCCcore/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/include/stddef.h'
  '/sw/installed/GCCcore/8.2.0/lib/gcc/x86_64-pc-linux-gnu/8.2.0/include/stdarg.h'
Target //tensorflow/tools/pip_package:build_pip_package failed to build
```

**Provide the exact sequence of commands / steps that you executed before running into the problem**

- `TF_NEED_CUDA=1 ./configure`
- `export TF_MKL_DOWNLOAD=1 &&   bazel --output_base=/tmp/easybuild-tmp/eb-5QGVSJ/tmpMZolLj-bazel-build --install_base=/tmp/easybuild-tmp/eb-5QGVSJ/tmpMZolLj-bazel-build/inst_base --output_user_root=/tmp/easybuild-tmp/eb-5QGVSJ/tmp514aJM-user_root build --compilation_mode=opt --config=opt --subcommands --verbose_failures --jobs=24 --action_env=PYTHONPATH --action_env=EBPYTHONPREFIXES --distinct_host_configuration=false  --config=mkl //tensorflow/tools/pip_package:build_pip_package`

Note that the installation is triggered via EasyBuild

**Any other info / logs**
[command.log](https://github.com/tensorflow/tensorflow/files/3804323/command.log)

**Comment by Flamefire** [comment]
This seems to happen only when CUDA is requested (TF_NEED_CUDA=1` or `--config=cuda`)

**Comment by AmericanEnglish** [comment]
I don't have any fix for this but I have also been struggling with getting it to compile from source with similar errors such as:
```
ERROR: /home/user/.tmp/external/zlib_archive/BUILD.bazel:5:1: undeclared inclusion(s) in rule '@zlib_archive//:zlib':
this rule is missing dependency declarations for the following files included by 'external/zlib_archive/uncompr.c':
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/stddef.h'
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include-fixed/limits.h'
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include-fixed/syslimits.h'
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/stdarg.h'
Target //tensorflow/tools/pip_package:build_pip_package failed to build
INFO: Elapsed time: 12.929s, Critical Path: 4.14s
```
or even
```
ERROR: /home/user/tests/tensorflowTest/fromsource/tensorflow/tensorflow/lite/c/BUILD:6:1: undeclared inclusion(s) in rule '//tensorflow/lite/c:c_api_internal':
this rule is missing dependency declarations for the following files included by 'tensorflow/lite/c/c_api_internal.c':
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/stdbool.h'
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/stddef.h'
  '/usr/ebuild/software/GCCcore/9.2.0/lib/gcc/x86_64-pc-linux-gnu/9.2.0/include/stdint.h'
Target //tensorflow/tools/pip_package:build_pip_package failed to build                                                                                                                                                                      INFO: Elapsed time: 13.962s, Critical Path: 6.44s
INFO: 3 processes: 3 local.
FAILED: Build did NOT complete successfully
```
I haven't been able to get anything to work despite editing a myriad of files, rerunning configure, clearing out the bazel cache, and recloning fresh.

**Comment by boegel** [comment]
I was able to dance around this issue using the following patch: https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/t/TensorFlow/TensorFlow-1.14.0_fix-cuda-build.patch .

If you're using EasyBuild: note that a sufficiently recent `CUDA` version is required by TensorFlow 2.0.0 (the one included in `fosscuda/2019a` is too old)

**Comment by Flamefire** [comment]
This patch does indeed work for us. It is up to the maintainers to decide whether this is a Bazel or a Tensorflow bug, I opened an issue there too. It seems the core issue is that `g++ -E -v -` returns resolved paths but the dependency checker does use plain/unresolved paths and Bazel does not try to resolve them there.

**Comment by AmericanEnglish** [comment]
I would to confirm that this patch works and thank @boegel for the link to the patch. I wrote a small python script to make changes to the highlighted file in my automatic installer for tensorflow 2.0 . 

Here's hoping to a more official fix from the maintainers of either bazel or tensorflow.

I still run into an issue where bazel is linking against the system libraries instead of the ones that come with gcc.
```
ERROR: /home/users/tests/tensorflowTest/fromsource/tensorflow/tensorflow/lite/python/interpreter_wrapper/BUILD:60:1: SWIGing tensorflow/lite/python/interpreter_wrapper/interpreter_wrapper.i failed (Exit 1)
bazel-out/host/bin/external/swig/swig: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by bazel-out/host/bin/external/swig/swig)
Target //tensorflow/tools/pip_package:build_pip_package failed to build
```

I'll open an issue on bazel since this seems beyond the current ticket.

**Comment by Flamefire** [comment]
The proposed patch does not work in some circumstances such as when using compiler wrappes like ccache as that throws of the relation between compiler and include paths. See https://github.com/bazelbuild/bazel/issues/10167

So I guess this must be fixed in Bazel.

**However**: Strangely this issue does not occur when not using the CUDA build, so for "normal" GCC it must be able to handle the include paths on symlinks. Most likely the creation of a derived toolchain is wrong/outdated? I've seen some discussion about handling "system" include paths internally in Bazel making this handling not required. But I don't understand enough about the Bazel internals to propose a change.

**Comment by preethivenkatesh** [comment]
I'm a little confused. so the issue is with Bazel not being able to resolve the paths in general or particularly when using mkl flag?

**Comment by Flamefire** [comment]
This has nothing to do with MKL, it is Bazel generally not able to match resolved vs unresolved paths, see https://github.com/tensorflow/tensorflow/issues/33975#issuecomment-549397196 and https://github.com/bazelbuild/bazel/issues/10167: It compares unresolved symlinks to resolved paths reported by gcc. @ymodak You might want to fix the tags and assignment.

A potential patch to TensorFlow is https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/t/TensorFlow/TensorFlow-2.1.0_fix-cuda-build.patch which works so far for us.

**Comment by amahendrakar** [comment]
@Flamefire,
Is this still an issue?

Could you please update TensorFlow to the latest stable version v.2.4.1 and let us know if you are facing the same error. Thanks!

**Comment by Flamefire** [comment]
Yes, even for 2.4.1 we still need to patch using https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/t/TensorFlow/TensorFlow-2.1.0_fix-cuda-build.patch

**Comment by r4nt** [comment]
We pass -no-canonical-prefixes and -no-canonical-system-headers so gcc does not resolve symlinks. If gcc still resolves symlinks, I'd be curious what happens.
Can somebody create a Dockerfile that I could use to debug that ends with this error?

**Comment by Flamefire** [comment]
> We pass -no-canonical-prefixes and -no-canonical-system-headers so gcc does not resolve symlinks.

Where? Not in https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/cuda_configure.bzl#L268 as far as I can tell

**Comment by Saduf2019** [comment]
@Flamefire 
Is this still an issue in latest tf version, please confirm and move to close status if it is not an issue.

**Comment by Flamefire** [comment]
Yes this is still an issue

**Comment by mohantym** [comment]
Hi @Flamefire ! Can you try [latest document instructions ](https://www.tensorflow.org/install/source)with GCC 7.3.2 , Bazel 3.7.2 and TF 2.7 and let us know?

**Comment by Flamefire** [comment]
Hi @mohantym I just got around testing this with the latest Bazel and TF but the issue still persists.

Possible solutions is our long-term patch at https://github.com/easybuilders/easybuild-easyconfigs/blob/439c0f2cb41bcb8f11071457550a2786eabe76a8/easybuild/easyconfigs/t/TensorFlow/TensorFlow-2.1.0_fix-cuda-build.patch or https://github.com/tensorflow/tensorflow/pull/56360

**Comment by mohantym** [comment]
Ok @Flamefire ! 
Thanks for the update. Re-opening as this still persists and is linked with PR [#56360](https://github.com/tensorflow/tensorflow/pull/56360)

**Comment by mohantym** [comment]
Ok @Flamefire ! 
Thanks for the update. Re-opening as this still persists and is linked with PR [#56360] . (https://github.com/tensorflow/tensorflow/pull/56360)

**Comment by njzjz** [comment]
This issue can be closed as #56360 has been merged.

**Comment by Flamefire** [comment]
@mohantym @njzjz Unfortunately the issue has only been partially solved. It recently reappeared in a slight variation with the recent TensorFlow.

The issue occurs when the compiler (e.g. `gcc`) is on a symlinked path. E.g. `/opt/gcc-current/bin/gcc` -> `/opt/gcc-11/bin/gcc`

https://github.com/tensorflow/tensorflow/pull/56360 fixes the issue when the compiler is directly used, i.e. `GCC_HOST_COMPILER_PATH=/opt/gcc-current/bin/gcc`

However you may have a compiler-wrapper in use, e.g.:
```
$ cat /opt/ccache/gcc
#!/usr/bin/bash
# ...
/opt/gcc-current/bin/gcc "$@"
```

#56360 doesn't fix that case as it will resolve `GCC_HOST_COMPILER_PATH=/opt/ccache/gcc` but not the contained symlink `/opt/gcc-current/bin/gcc` and later error out with undeclared inclusion false positives.
